pubsub/pubsub.go

```diff
@@ -562,20 +562,37 @@ func (s *Subscription) Receive(ctx context.Context) (_ *Message, err error) {
 			// log.Printf("BATCH SIZE %d", batchSize)
 
 			go func() {
+				defer func() {
+					close(s.waitc)
+					s.waitc = nil
+				}()
 				if s.preReceiveBatchHook != nil {
 					s.preReceiveBatchHook(batchSize)
 				}
-				msgs, err := s.getNextBatch(batchSize)
-				s.mu.Lock()
-				defer s.mu.Unlock()
-				if err != nil {
-					// Non-retryable error from ReceiveBatch -> permanent error.
-					s.err = err
-				} else if len(msgs) > 0 {
-					s.q = append(s.q, msgs...)
+				resultChannel := s.getNextBatch(batchSize)
+				for {
+					select {
+					case msgs, ok := <-resultChannel.msgs:
+						if !ok {
```

Comment:
Can't this lose errors?

I.e., if `getNextBatch` gets an error, and writes it to resultChannel.err, and then immediately closes resultChannel.msgs, I think this `select` could end up here, and we'd never receive the error.

Instead of returning a struct that contains two channels, I think it might be better to return a single channel, which delivers structs (where the struct can either have messages or an error).

Reply:
I think there might be a bit more to it...

So, given that `g.Wait` in `getNextBatch` will return only after each goroutine has exited, it means that any messages sent, will certainly happen before `Wait` returns.

So when `Wait` returns, if there was an error it will be written to the error channel which is **not buffered** . So, this will block the execution until it is consumed. So it won't lose any errors since it won't move to close the channel unless the error gets consumed.

But you may be right about the flow being simpler returning a single channel. I will try this out